### PR TITLE
Feature: Discover product dependencies in a lazy, pluggable way

### PR DESCRIPTION
**Pull request by ferozco**
## After this PR

==COMMIT_MSG==
Discover product dependencies in a lazy, pluggable way
==COMMIT_MSG==

### CODE REVIEW

**Comment by ferozco** [comment]
Pending #1159

File: gradle-recommended-product-dependencies/src/main/java/com/palantir/gradle/dist/ConfigureProductDependenciesTask.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
```
Comment (carterkozak) [inline gradle-recommended-product-dependencies/src/main/java/com/palantir/gradle/dist/ConfigureProductDependenciesTask.java]:
```suggestion
 * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
```

File: gradle-recommended-product-dependencies/src/main/java/com/palantir/gradle/dist/CompileRecommendedProductDependencies.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Set;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.provider.SetProperty;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+
+public abstract class CompileRecommendedProductDependencies extends DefaultTask {
+    static final ObjectMapper MAPPER = new ObjectMapper();
```
Comment (carterkozak) [inline gradle-recommended-product-dependencies/src/main/java/com/palantir/gradle/dist/CompileRecommendedProductDependencies.java:30]: Do we need to worry about this ObjectMapper being configured differently from `ObjectMappers.jsonMapper`?

Comment (ferozco) [reply gradle-recommended-product-dependencies/src/main/java/com/palantir/gradle/dist/CompileRecommendedProductDependencies.java:30]: good catch, we should use the shared `ObjectMappers` mapper

Comment (ferozco) [reply gradle-recommended-product-dependencies/src/main/java/com/palantir/gradle/dist/CompileRecommendedProductDependencies.java:30]: err actually they're in separate packages. I don't think it ends up mattering but if it did we have tests that exercise that the JSON produced by this task are consumable in the main plugin

**Review by carterkozak — APPROVED** [review]
lgtm

**Pull request merged**